### PR TITLE
fix(errors): return actionable errors from API

### PR DESCRIPTION
**Pull request by Joshua-Anderson**
This tries to strike a balance between supporting every possible error, which is infeasible because of how many (undocumented) errors exist, and supporting the most useful errors.

The most useful errors have their own type, and the less common errors return a unknown error.

### CODE REVIEW

**Comment by Joshua-Anderson** [comment]
I also intend to add a `func IsUnkownError(err error) bool` so that it is easy to check to see if the error was not recognized.

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	defer func() {
+		io.Copy(ioutil.Discard, res.Body)
+		res.Body.Close()
+	}()
+
+	switch res.StatusCode {
+	case 400:
+		bodyMap := make(map[string]interface{})
+		if err := json.NewDecoder(res.Body).Decode(&bodyMap); err != nil {
+			return ErrParsing
```
Comment (bacongobbler) [inline errors.go]: perhaps we should also return the response body so the user can determine the error on their own?

File: apps/apps.go [hunk 2]
```diff
@@ -56,14 +58,18 @@ func New(c *deis.Client, id string) (api.App, error) {
 		}
 	}
 
-	resBody, err := c.BasicRequest("POST", "/v2/apps/", body)
-
+	res, err := c.Request("POST", "/v2/apps/", body)
 	if err != nil {
 		return api.App{}, err
 	}
+	// Fix json.Decoder bug in <go1.7
```
Comment (mboersma) [inline apps/apps.go]: Maybe this should be `// TODO: Remove when json.Decoder bug is fixed (go >= 1.7)` ? It's unfortunate that this has to be repeated everywhere `Decode` is used.

Also do you have a reference for this json.Decoder issue? Just curious.

Comment (Joshua-Anderson) [inline apps/apps.go]: It was caught in google's github-api repo: https://github.com/google/go-github/pull/317

I'll make a PR to reformat the warning to what you suggested everywhere.

**Comment by Joshua-Anderson** [comment]
This is no longer WIP. @helgi you had some concerns about my approach, could you please review?

**Comment by Joshua-Anderson** [comment]
@bacongobbler You accidentally LGTMed this twice ❤️! I removed you second LGTM so somebody else can review it.

**Comment by bacongobbler** [comment]
whoops :P

**Comment by helgi** [comment]
I'd rather want to fix the API where possible with more identifable information than to parse things like this but good enough for now. An improvement for sure

**Pull request merged**